Cloud Vision Document OCR Template

dzou opened the pull request:
Implement template for running OCR on documents stored on GCS through Cloud Vision APIs.

Note: The main design change from past discussions is in `runOcrForDocument` which will produce one JSON output file for each page of the document (i.e. it is paginated). I decided to do this because the API [actually forces paginated output (default is 20 pages of document per JSON output)](https://cloud.google.com/vision/docs/reference/rest/v1p2beta1/files/asyncBatchAnnotate#OutputConfig). This was something that we overlooked in our team discussions. In my discussion with Vision team, they said that producing 1 json file per page of document was a common practice, so I shifted to this practice here.

elefeint commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java:
```diff
@@ -0,0 +1,250 @@
+			public void onSuccess(
+					AsyncBatchAnnotateFilesResponse asyncBatchAnnotateFilesResponse) {
+
+				String outputFolderUri = asyncBatchAnnotateFilesResponse.getResponsesList().get(0)
```
Is it possible to have ore than one response?

dzou replied:
Nope, this is safe - this is due to that 1 request per batch request that we discussed during our meetings. Only our code will be providing the response object here.

dmitry-s commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,99 @@
+	 *     parsed
+	 */
+	public TextAnnotation getPage(int pageNumber) throws InvalidProtocolBufferException {
+		Blob pageBlob = this.pageBlobs.get(pageNumber);
```
maybe check for `pageNumber < getPageCount()` and throw appropriate exception?

dzou replied:
Done.

elefeint commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java:
```diff
@@ -0,0 +1,250 @@
+
+		AnnotateFileResponse annotateFileResponse = annotateFileResponseBuilder.build();
+
+		return annotateFileResponse.getResponses(0).getFullTextAnnotation();
```
Is it possible to have more than 1 response?

dzou replied:
Nope, this is also safe for the same reason as above. Only our code will be able to provide the response object (with 1 response inside).

dmitry-s commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,99 @@
+
+	public DocumentOcrResultSet(List<Blob> pages, Storage storageClient) {
+		this.pageBlobs = pages;
+		this.storageClient = storageClient;
```
Why do we have `storageClient` here? It seems it is not used anywhere.

dzou replied:
Nice catch. Done.

elefeint commented on spring-cloud-gcp-vision/src/test/java/org/springframework/cloud/gcp/vision/it/DocumentOcrTemplateIntegrationTests.java:
```diff
@@ -0,0 +1,94 @@
+import static org.assertj.core.api.Assumptions.assumeThat;
+
+@RunWith(SpringRunner.class)
+@ContextConfiguration(classes = { VisionTestConfiguration.class })
```
Is it possible to rely on real configuration? It's nice when integration tests verify the way starter will be used in real life.

dzou replied:
Agreed, unfortunately I couldn't quite get this to work. Currently, the autoconfiguration project depends on the Vision project ([to provide beans of Vision templates](https://github.com/spring-cloud/spring-cloud-gcp/blob/master/spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/vision/CloudVisionAutoConfiguration.java#L85)). If I reference the real configuration here, I would have to make the vision project depend on autoconfiguration project which produces a circular dependency.

Do you have any suggestions? Maybe I am missing something with how Spring dependency injection works... In the Guice world I would only have the vision autoconfiguration provide `ImageAnnotatorClient` and be able to delete the autoconfiguration's dependency on the vision module. However, it seems like in Spring, I must declare the beans for the templates explicitly which forces the dependency between autoconfiguration and vision.

elefeint replied:
Makes sense, thanks! This way we exercise both, the pure template usage, and the template+autoconfig usage in sample test.

dzou commented on spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/vision/CloudVisionAutoConfiguration.java:
```diff
@@ -86,4 +88,11 @@ public CloudVisionTemplate cloudVisionTemplate(ImageAnnotatorClient imageAnnotat
+	@Bean
+	@ConditionalOnMissingBean
+	public DocumentOcrTemplate documentOcrTemplate(
+			ImageAnnotatorClient imageAnnotatorClient, Storage storage) {
```
Yup, it's provided above!

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,99 @@
+	 */
+	public TextAnnotation getPage(int pageNumber) throws InvalidProtocolBufferException {
+		Blob pageBlob = this.pageBlobs.get(pageNumber);
+		return DocumentOcrTemplate.parseJsonBlob(pageBlob);
```
Do you mean the whole method or specifically `DocumentOcrTemplate.parseJsonBlob`?

Open to suggestions - to give some more context, the `parseJsonBlob` method converts a "blob" file on GCS to a TextAnnotation proto. It's package-private static helper and shared between this class and DocumentOcrTemplate.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,99 @@
+			@Override
+			public TextAnnotation next() {
+				try {
+					TextAnnotation result = getPage(currentPage);
```
Done.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java:
```diff
@@ -0,0 +1,250 @@
+				.setUri(document.uriString())
+				.build();
+
+		Blob documentBlob = this.storage.get(
```
Done.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java:
```diff
@@ -0,0 +1,250 @@
+				.build();
+
+		GcsDestination gcsDestination = GcsDestination.newBuilder()
+				.setUri(outputFilePathPrefix.uriString())
```
The outputFilePathPrefix as very little restrictions (it can describe a file/folder/bucket) - in this case we can skip it because we can defer the error handling to the Vision API which will do validation on the URI.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java:
```diff
@@ -0,0 +1,250 @@
+
+		OutputConfig outputConfig = OutputConfig.newBuilder()
+				.setGcsDestination(gcsDestination)
+				.setBatchSize(1)
```
Done. This field means it will produce 1 JSON file per page in the document. If unset, this will [default to 20 pages per JSON file](https://cloud.google.com/vision/docs/reference/rpc/google.cloud.vision.v1#google.cloud.vision.v1.OutputConfig). And the range of values is 1-100.

I set it to 1 (aka 1-page-per-json-file) because it seemed like an intuitive default for someone who doesn't want to think about it. However upon reinspection it might make sense to make it configurable by the user ? (I suppose someone with a 1000-page pdf wouldn't want to output 1000 json files.)

elefeint replied:
Configurable would be nice as a follow-up non-urgent PR.

dzou replied:
I ultimately changed it to configurable b/c I realized this is an important feature, i.e. batching seems important because it will reduce the amount of I/O needed for large documents.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java:
```diff
@@ -0,0 +1,250 @@
+				.setBatchSize(1)
+				.build();
+
+		Feature feature = Feature.newBuilder()
```
Done.

dzou commented:
Hi guys, I've adjusted this PR based on your comments. Also modified it such that it can support user specifying how many document pages to include in each JSON OCR output file. PTAL.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java:
```diff
@@ -60,12 +61,17 @@
 
 	private final Storage storage;
 
+	@Autowired
```
Done

dzou commented:
Made changes as discussed in our meeting; please take another look.

meltsufin commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,198 @@
+		}
+
+		Map.Entry<Integer, OcrPageRange> floorEntry = this.ocrPageRanges.floorEntry(pageNumber);
+		OcrPageRange pageRange = floorEntry.getValue();
```
Why don't you check the `pageNumber` against the `pageRange` here, before parsing?

dzou replied:
Done.

meltsufin commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,198 @@
+
+		Map.Entry<Integer, OcrPageRange> floorEntry = this.ocrPageRanges.floorEntry(pageNumber);
+		OcrPageRange pageRange = floorEntry.getValue();
+		List<TextAnnotation> documentPages = DocumentOcrTemplate.parseJsonBlob(pageRange.blob);
```
I think we discussed caching the parsed json blobs. Can we do that?

dzou replied:
Done.

meltsufin commented on spring-cloud-gcp-starters/spring-cloud-gcp-starter-vision/pom.xml:
```diff
@@ -27,5 +27,10 @@
+
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-gcp-storage</artifactId>
```
Do you need this dep here at all? Since the starter doesn't have any code, I don't think it makes sense to have any "optional" dependencies in it.

dzou replied:
Removed.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,198 @@
+ */
+public class DocumentOcrResultSet {
+
+	private static final Pattern OUTPUT_PAGE_PATTERN = Pattern.compile("output-(\\d+)-to-(\\d+)\\.json");
```
For now I don't see a reason why the user would want to rename the output files after they have been generated, so I think it makes sense to just keep it simple for now.

dzou commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,198 @@
+
+	private static final Pattern OUTPUT_PAGE_PATTERN = Pattern.compile("output-(\\d+)-to-(\\d+)\\.json");
+
+	private final TreeMap<Integer, OcrPageRange> ocrPageRanges;
```
Tried to, however I am using `.floorEntry(...)` which appears to only be in TreeMap and not in SortedMap.

meltsufin commented on spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrResultSet.java:
```diff
@@ -0,0 +1,225 @@
+
+	private final TreeMap<Integer, OcrPageRange> ocrPageRanges;
+
+	private final TreeMap<Integer, List<TextAnnotation>> documentPagesCache;
```
Can we instead just use one simple `HashMap<Integer, OcrPageRange>`, where there is an entry for every single page? The `getTextAnnotations()` can move to the `OcrPageRange` class, and caching can very easily be done there. I think it would simplify your code quiet a bit.

dzou replied:
Partially done. I moved the caching into OcrPageRange which simplified the caching code.

However, I didn't switch from TreeMap to HashMap. I don't think using TreeMap here is that much more complex that HashMap, so might as well do things efficiently. For users processing with JSON `batchSize = 100` we would reduce the number of entries stored in map by factor of 100.

meltsufin commented on spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/vision/CloudVisionProperties.java:
```diff
@@ -35,7 +35,27 @@
 
+	private int executorThreadsCount = 1;
+
+	private int jsonOutputBatchSize = 20;
```
I think there is a javadoc requirement for these? Can you look at other `...Properties` classes for an example?

dzou replied:
Gotcha, will add these with the docs. As reminder to add javadoc to properties, I created https://github.com/spring-cloud/spring-cloud-gcp/issues/1548 to track.

elefeint commented on spring-cloud-gcp-vision/pom.xml:
```diff
@@ -24,5 +24,18 @@
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-gcp-storage</artifactId>
+			<optional>true</optional>
```
Is it optional? VisionTemplate depends on Storage in all cases.

meltsufin replied:
Just for OCR, I believe.

dzou replied:
Yeah I wasn't 100% sure to be honest; it's what I had before but Mike suggested it could be marked optional since only the OCR template uses storage but not the general one. I figure once I merge this with my sample I'll be able to figure it out.
